wsdl-core/src/main/java/io/ballerina/xsd/core/visitor/VisitorUtils.java

```diff
@@ -0,0 +1,233 @@
+        if (minOccurrence != null) {
+            if (minOccurrence.getNodeValue().equals(ZERO)) {
+                builder.append(QUESTION_MARK);
+            }
+        }
+    }
+
+    public static void handleMaxOccurrences(Node node, StringBuilder builder) {
+        Node maxOccurrence = node.getAttributes().getNamedItem(MAX_OCCURS);
+        if (maxOccurrence != null) {
+            String maxValue = maxOccurrence.getNodeValue();
+            builder.append(maxValue.equals(UNBOUNDED)
+                    ? OPEN_SQUARE_BRACKET + CLOSE_SQUARE_BRACKET + WHITESPACE : !maxValue.equals(ONE)
+                    ? OPEN_SQUARE_BRACKET + EMPTY_STRING + CLOSE_SQUARE_BRACKET + WHITESPACE : EMPTY_STRING
+            );
+        }
+    }
+
+    public static String deriveType(Node node) {
+        String derivedType = (node != null)
+                ? node.getNodeValue().contains(COLON)
+                ? node.getNodeValue().substring(node.getNodeValue().indexOf(COLON) + 1)
+                : node.getNodeValue()
+                : STRING;
+        return typeGenerator(derivedType);
+    }
+
+    public static String typeGenerator(String typeName) {
+        switch (typeName) {
+            case TIME, DATE_TIME, DATE, G_YEAR_MONTH, G_YEAR, STRING, LANGUAGE, DURATION, ANY_URI -> {
+                return STRING;
+            }
+            case INTEGER, LONG, NEGATIVE_INTEGER, NON_POSITIVE_INTEGER, POSITIVE_INTEGER, SHORT,
+                    UNSIGNED_LONG, UNSIGNED_INT, UNSIGNED_SHORT, UNSIGNED_BYTE, INT, NON_NEGATIVE_INTEGER -> {
+                return INT;
+            }
+            case BASE64_BINARY -> {
+                return BYTE_ARRAY;
+            }
+            default -> {
+                return typeName;
+            }
+        }
+    }
+
+    public static boolean isSimpleType(String type) {
+        String typeName = type.contains(COLON) ? type.substring(type.indexOf(COLON) + 1) : type;
+        String[] simpleTypes = {
+                TIME, DATE_TIME, DATE, G_YEAR_MONTH, G_YEAR, STRING, LANGUAGE, DURATION,
+                INTEGER, LONG, NEGATIVE_INTEGER, NON_POSITIVE_INTEGER, POSITIVE_INTEGER, SHORT,
+                UNSIGNED_LONG, UNSIGNED_INT, UNSIGNED_SHORT,
+                UNSIGNED_BYTE, INT, BASE64_BINARY, BOOLEAN, FLOAT, DOUBLE, DECIMAL, ANY_URI, NON_NEGATIVE_INTEGER
+        };
+        return Arrays.stream(simpleTypes).toList().contains(typeName);
+    }
+
+    public static Iterable<Node> asIterable(NodeList nodeList) {
+        return () -> new Iterator<>() {
+            private int index = 0;
+
+            @Override
+            public boolean hasNext() {
+                return index < nodeList.getLength();
+            }
+
+            @Override
+            public Node next() {
+                if (!hasNext()) {
+                    throw new NoSuchElementException();
+                }
+                return nodeList.item(index++);
+            }
+        };
+    }
```

Comment:
Isn't NodeList already an Iterable? Do we need this function ?

Reply:
Yes, but the `item(index)` API in `NodeList` has index-based iteration, which can make the code look less clean and readable